File: Data Structures & Algorithms/reorder-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {

ListNode* findCenterAndPrev(ListNode* head){
    if(!head || !head->next) return head ;
     ListNode* prev = NULL ;
        auto slow = head ;
        auto fast = head ;
        while(fast && fast->next){
            prev = slow ;
            slow = slow->next ;
            fast = fast->next->next ;
        }

        return prev ;
}

ListNode* reverse(ListNode* head){
    if(!head || !head->next) return head;
    auto temp = head->next ;
    head->next = NULL ;
    while(temp){
        auto temp1=temp->next ;
        temp->next = head; 
        head = temp ;
        temp = temp1 ;
    }

    return head ;
}

public:
    void reorderList(ListNode* head) {
        ListNode* prev = findCenterAndPrev(head) ;
        auto head1 = prev->next ;
        prev->next = NULL ;
        auto head2 = reverse(head1) ;
        prev = head ;
        while(head2 && prev && prev->next){
            auto temp1 = prev->next ;
            prev->next = head2 ;
            head2 = head2->next ;
            prev->next->next = temp1 ;
            prev = temp1 ;
        }
        if(head2) prev->next = head2 ;
    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        def add_heads(prev_val, head1, head2):

            if not head1 and not head2:
                if prev_val:
                    return ListNode(prev_val)
                return None

            if not head1:
                new_val = head2.val + prev_val
                if new_val > 9:
                    head2.val = new_val % 10
                    head2.next = add_heads(1, None, head2.next)
                else:
                    head2.val = new_val
                return head2

            if not head2:
                new_val = head1.val + prev_val
                if new_val > 9:
                    head1.val = new_val % 10
                    head1.next = add_heads(1, head1.next, None)
                else:
                    head1.val = new_val
                return head1

            new_val = head1.val + head2.val + prev_val

            if new_val > 9:
                head1.val = new_val % 10
                head1.next = add_heads(1, head1.next, head2.next)
            else:
                head1.val = new_val
                head1.next = add_heads(0, head1.next, head2.next)

            return head1

        return add_heads(0,l1,l2)